llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -756,24 +756,56 @@ processInputModule(std::unique_ptr<Module> M) {
       module_split::getSplitterByMode(module_split::ModuleDesc{std::move(M)},
                                       SplitMode, IROutputOnly,
                                       EmitOnlyKernelsAsEntryPoints);
-  const bool SplitByScope = ScopedSplitter->totalSplits() > 1;
-  Modified |= SplitByScope;
 
+  SmallVector<module_split::ModuleDesc, 8> TopLevelModules;
+  bool SplitByOptionalFeatures = false;
+
+  // FIXME: this check should be performed on all split levels
   if (DeviceGlobals)
     ScopedSplitter->verifyNoCrossModuleDeviceGlobalUsage();
 
+  while (ScopedSplitter->hasMoreSplits()) {
+    module_split::ModuleDesc MD = ScopedSplitter->nextSplit();
+
+    if (IROutputOnly || SplitMode == module_split::SPLIT_NONE) {
+      // We can't perform any kind of split
+      TopLevelModules.emplace_back(std::move(MD));
+      continue;
+    }
+
+    std::unique_ptr<module_split::ModuleSplitterBase> OptionalFeaturesSplitter =
+        module_split::getSplitterByOptionalFeatures(
+            std::move(MD), EmitOnlyKernelsAsEntryPoints);
+
+    // Here we perform second-level splitting based on device-specific
+    // features used/declared in entry points.
+    // This step is mandatory, because it is required for functional
+    // correctness, i.e. to prevent speculative compilation of kernels that use
+    // optional features on a HW which doesn't support them.
+    while (OptionalFeaturesSplitter->hasMoreSplits()) {
+      TopLevelModules.emplace_back(OptionalFeaturesSplitter->nextSplit());
+    }
+
+    SplitByOptionalFeatures |= OptionalFeaturesSplitter->totalSplits() > 1;
+  }
+
+  const bool SplitByScope = ScopedSplitter->totalSplits() > 1;
+  Modified |= SplitByScope;
+  Modified |= SplitByOptionalFeatures;
+
   // TODO this nested splitting scheme will not scale well when other split
   // "dimensions" will be added. Some infra/"split manager" needs to be
   // implemented in this case - e.g. all needed splitters are registered, then
   // split manager applies them in the order added and runs needed tforms on the
   // "leaf" ModuleDesc's resulted from splitting. Some bookkeeping is needed for
   // ESIMD splitter to link back needed modules.
 
-  // Proceed with top-level splitting.
-  while (ScopedSplitter->hasMoreSplits()) {
-    module_split::ModuleDesc MDesc = ScopedSplitter->nextSplit();
+  // Based on results from a top-level splitting, we perform some lower-level
```

Comment:
Not a native speaker, so I might be wrong here.

```suggestion
  // Based on results from the top-level splitting, we perform some lower-level
```

Reply:
Applied in 0cc45b2c1a04551249551bf104bd99b559a55d83